x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/SetSingleNodeAllocateStep.java

```diff
@@ -44,15 +46,23 @@ public SetSingleNodeAllocateStep(StepKey key, StepKey nextStepKey, Client client
 
     @Override
     public void performAction(IndexMetaData indexMetaData, ClusterState clusterState, ClusterStateObserver observer, Listener listener) {
-        RoutingAllocation allocation = new RoutingAllocation(ALLOCATION_DECIDERS, clusterState.getRoutingNodes(), clusterState, null,
+        final RoutingNodes routingNodes = clusterState.getRoutingNodes();
+        RoutingAllocation allocation = new RoutingAllocation(ALLOCATION_DECIDERS, routingNodes, clusterState, null,
                 System.nanoTime());
         List<String> validNodeIds = new ArrayList<>();
         Optional<ShardRouting> anyShard = clusterState.getRoutingTable().allShards(indexMetaData.getIndex().getName()).stream().findAny();
         if (anyShard.isPresent()) {
+            // Find the highest node version in the cluster - shards can't be allocated to a node running an older version than they're
+            // currently on, so we should only pick one of the newest version nodes to reallocate to
+            Version highestNodeVersion = indexMetaData.getCreationVersion();
```

Comment:
I think instead of the version the index was created with, we want the combined versions of the *nodes* that the index currently resides on. That's what we use for the shard's decision criteria in `NodeVersionAllocationDecider`